From 9522aa00021cd61e2b994a7ed1b5c9767442f41a Mon Sep 17 00:00:00 2001
From: Tommy Lin <lin.tommysc@inventec.com>
Date: Tue, 22 Feb 2022 05:24:56 +0000
Subject: [PATCH 2/3] Implement OperationalStatus dbus interface

Symptom/Reason:
    - ProcessorSummary state should be "Enable" and count more than 0.

Root Cause:
    - OperationalStatus dbus interface is not implemented

Solution/Change:
    [smbios-mdr]
      - Implement OperationalStatus dbus interface

Entry Test:
    Get redfish system info
      - curl -X GET https://${bmc}/redfish/v1/Systems/system
---
 include/cpu.hpp | 6 ++++--
 src/cpu.cpp     | 2 ++
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/include/cpu.hpp b/include/cpu.hpp
index 6328393..d857330 100644
--- a/include/cpu.hpp
+++ b/include/cpu.hpp
@@ -24,6 +24,7 @@
 #include <xyz/openbmc_project/Inventory/Decorator/Revision/server.hpp>
 #include <xyz/openbmc_project/Inventory/Item/Cpu/server.hpp>
 #include <xyz/openbmc_project/Inventory/Item/server.hpp>
+#include <xyz/openbmc_project/State/Decorator/OperationalStatus/server.hpp>
 
 namespace phosphor
 {
@@ -43,6 +44,7 @@ using processor = sdbusplus::server::xyz::openbmc_project::inventory::item::Cpu;
 using Item = sdbusplus::server::xyz::openbmc_project::inventory::Item;
 using association =
     sdbusplus::server::xyz::openbmc_project::association::Definitions;
+using func = sdbusplus::xyz::openbmc_project::State::Decorator::server::OperationalStatus;
 
 // Definition follow smbios spec DSP0134 3.0.0
 static const std::map<uint8_t, const char*> familyTable = {
@@ -113,7 +115,7 @@ static const std::array<std::optional<processor::Capability>, 16>
 
 class Cpu :
     sdbusplus::server::object_t<processor, asset, location, connector, rev,
-                                Item, association>
+                                Item, association, func>
 {
   public:
     Cpu() = delete;
@@ -126,7 +128,7 @@ class Cpu :
     Cpu(sdbusplus::bus_t& bus, const std::string& objPath, const uint8_t& cpuId,
         uint8_t* smbiosTableStorage, const std::string& motherboard) :
         sdbusplus::server::object_t<processor, asset, location, connector, rev,
-                                    Item, association>(bus, objPath.c_str()),
+                                    Item, association, func>(bus, objPath.c_str()),
         cpuNum(cpuId), storage(smbiosTableStorage), motherboardPath(motherboard)
     {
         infoUpdate();
diff --git a/src/cpu.cpp b/src/cpu.cpp
index 7c10a09..c5672ec 100644
--- a/src/cpu.cpp
+++ b/src/cpu.cpp
@@ -152,6 +152,7 @@ void Cpu::infoUpdate(void)
     {
         // Don't attempt to fill in any other details if the CPU is not present.
         present(false);
+        functional(false);
         return;
     }
     present(true);
@@ -230,6 +231,7 @@ void Cpu::infoUpdate(void)
     }
 
     characteristics(cpuInfo->characteristics); // offset 26h
+    functional(true);
 
     if (!motherboardPath.empty())
     {
-- 
2.17.1

